### PR TITLE
[BugFix] fix race condition issue about persistent index's l0

### PR DESCRIPTION
**Pull request by luohaha**
## Why I'm doing:
In apply thread, persistent index will be modified and `_l0` can be destroy and re-create. At the same time, another thread may visit `_l0` by `PersistentIndex::memory_usage`. 
There will be race condition issue here.

## What I'm doing:
Add atomic val `_memory_usage`, calculate it each time when persistent index been modified, and `PersistentIndex::memory_usage` is only need to return `_memory_usage`, which is thread safe.

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

**Comment by trueeyu** [comment]
Add test case

**Comment by trueeyu** [comment]
```
        std::unique_lock wrlock(_lock);
        _l1_vec.emplace_back(std::move(l1_st).value());
        _l1_merged_num.emplace_back(1);
    }
    // clear l0
    _l0->clear();   please check this code
```

**Comment by trueeyu** [comment]
```
    new_l1_vec.emplace_back(std::move(l1_st).value());
    new_l1_merged_num.emplace_back((merge_l1_end_idx - merge_l1_start_idx) * merge_num);
    {
        std::unique_lock wrlock(_lock);
        _l1_vec.swap(new_l1_vec);
        _l1_merged_num.swap(new_l1_merged_num);
    }
    _l0->clear();  //here
```

File: be/src/storage/persistent_index.cpp [hunk 1]
```diff
@@ -5053,7 +5050,6 @@ Status PersistentIndex::_load_by_loader(TabletLoader* loader) {
 
     // clear l1
     {
```
Comment (trueeyu) [inline be/src/storage/persistent_index.cpp:5052]: "{" can be removed?

Comment (luohaha) [reply be/src/storage/persistent_index.cpp]: Done

File: be/src/storage/persistent_index.h [hunk 2]
```diff
@@ -656,21 +656,7 @@ class PersistentIndex {
...
-        size_t memory_usage = _l0 ? _l0->memory_usage() : 0;
-        for (int i = 0; i < _l1_vec.size(); i++) {
-            memory_usage += _l1_vec[i]->memory_usage();
-        }
-        for (int i = 0; i < _l2_vec.size(); i++) {
-            memory_usage += _l2_vec[i]->memory_usage();
-        }
-        return memory_usage;
-    }
+    size_t memory_usage() const { return _memory_usage.load(); }
```
Comment (trueeyu) [inline be/src/storage/persistent_index.h:660]: will `PersistentIndex::load_from_tablet` update the mem_usage?

Comment (trueeyu) [inline be/src/storage/persistent_index.h:660]: Will `PersistentIndex::commit` update the mem_usage?

Comment (trueeyu) [reply be/src/storage/persistent_index.h:660]: PersistentIndex::insert / PersistentIndex::upser / PersistentIndex::erase

Comment (luohaha) [reply be/src/storage/persistent_index.h:660]: load_from_tablet will call `load` or `commit`. Already added.

Comment (luohaha) [reply be/src/storage/persistent_index.h:660]: Already added. In `_flush_advance_or_append_wal`

File: be/test/storage/primary_index_test.cpp [hunk 3]
```diff
@@ -353,7 +353,7 @@ PARALLEL_TEST(PrimaryIndexTest, test_composite_key) {
     PrimaryKeyEncoder::encode(*schema, *chunk, 0, chunk->num_rows(), pk_column.get());
 
     ASSERT_TRUE(pk_index->insert(0, 0, *pk_column).ok());
-    LOG(INFO) << "pk_index memory:" << pk_index->memory_usage() << " capacity:" << pk_index->capacity();
+    LOG(INFO) << "pk_index memory:" << pk_index->memory_usage();
```
Comment (trueeyu) [inline be/test/storage/primary_index_test.cpp:356]: why remove pk_index->capacity();?

Comment (luohaha) [reply be/test/storage/primary_index_test.cpp:356]: It's basically useless at the moment.

**Review by trueeyu — APPROVED** [review]

**Comment by trueeyu** [comment]
Will 2.5/3.0 have the problem?

**Review by decster — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**